### PR TITLE
Fallback to customized Product Archive template if no better Category or Tag templates are present

### LINKED ISSUE #5520: Edits to Product Archive template are not visible in the Product Category and Product Tag pages
I might be understanding this wrong, but I think that if Product Category and Product Tag templates haven't been customized and don't have a custom template set by the theme, but the Product Archive has been customized by the user, the Product Archive template should take priority.

## To reproduce
1. With WP 5.9 and a Twenty Twenty Two, go to Appearance > Editor > Templates.
2. Edit the Product Archive template.
3. Now, go to a product category or product tag page in the frontend (ie: `/product-category/clothing/`).
4. Notice changes are not applied in that page.

## Expected behavior
Any change in the Product Archive template should be applied to the Product Category and Product Tag pages if they hadn't been customized or the theme doesn't have a specific template for them.

## Screenshots
_Editor:_
imatge

_Frontend:_
imatge

### WordPress (please complete the following information):
* Core version: tested in WP 5.9 beta 4 and RC 1
* Gutenberg version (if installed): tested with GB 12.3.0 and disabled
* WooCommerce version: `trunk` and 6.1 RC1
* WooCommerce Blocks version: `trunk` and disabled

### LINKED ISSUE #42656: Edits to Product Archive template are not visible in the Product Category and Product Tag pages
I might be understanding this wrong, but I think that if Product Category and Product Tag templates haven't been customized and don't have a custom template set by the theme, but the Product Archive has been customized by the user, the Product Archive template should take priority.

## To reproduce
1. With WP 5.9 and a Twenty Twenty Two, go to Appearance > Editor > Templates.
2. Edit the Product Archive template.
3. Now, go to a product category or product tag page in the frontend (ie: `/product-category/clothing/`).
4. Notice changes are not applied in that page.

## Expected behavior
Any change in the Product Archive template should be applied to the Product Category and Product Tag pages if they hadn't been customized or the theme doesn't have a specific template for them.

## Screenshots
_Editor:_
imatge

_Frontend:_
imatge

### WordPress (please complete the following information):
* Core version: tested in WP 5.9 beta 4 and RC 1
* Gutenberg version (if installed): tested with GB 12.3.0 and disabled
* WooCommerce version: `trunk` and 6.1 RC1
* WooCommerce Blocks version: `trunk` and disabled

### PR DESCRIPTION
**Pull request by sunyatasattva**
If the user has customized the `archive-product` template and there is no better choice for the Category
or Tag, then the custom Product Archive is chosen.

The priority order is the following:

1. Custom Taxonomy Template
2. Taxonomy Template from the Theme
3. Custom Product Archive Template
4. Product Archive Template from the Theme

Fixes woocommerce/woocommerce#42656

P.S. Not sure about the labels here.

### Screenshots

Screen Shot 2022-01-13 at 23 38 57

#### Manual Testing

How to test the changes in this Pull Request:

1. Enable a theme with an `archive-product.html` but no taxonomy templates (or add your own to the theme).
2. Go to Appearance → Editor → Browse All Templates.
3. Notice that the Category and Tag templates are also available from the theme (previous behaviour).
4. Customize `archive-product` and save.
5. Notice that the Category and Tag templates now also appear as customized and are applied the same edits.

> **Note:** Even though they will have the customized blue dot next to them signaling to the user that some edit has been implemented, they will not appear at the top of the list because they have not received any direct customization. Because of the priority order above (template from theme takes prio over custom product archive), doing otherwise would make things messy. However, happy to talk about this in detail if required.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above

### Changelog

> Customizations to the theme Product Archive block template are applied to Category and Tag templates is no better alternative is available

### CODE REVIEW

**Comment by tjcafferkey** [comment]
Few things to note:
* If I enable a theme that does **not** have the `archive-product.html` template and gets them provided from Woo Blocks. The templates screen is missing Product Category and Product Tag templates.
* Customizations of the themes `archive-product.html` do not render on the frontend for product category/tag pages.
* Customizations of the themes `archive-product.html` template do not appear in the Site Editor for product category/tag templates.

Not sure if I'm doing anything wrong here?

File: src/BlockTemplatesController.php [hunk 1]
```diff
@@ -341,10 +341,22 @@ function ( $template ) use ( $template_slug ) {
 				continue;
 			}
 
-			// If the theme has an archive-product.html template, but not a taxonomy-product_cat.html template let's use the themes archive-product.html template.
+			// Category and tags template are eligible to fallback to the Product Archive if available.
 			if ( BlockTemplateUtils::template_is_eligible_for_product_archive_fallback( $template_slug ) ) {
-				$template_file = BlockTemplateUtils::get_theme_template_path( 'archive-product' );
-				$templates[]   = BlockTemplateUtils::create_new_block_template_object( $template_file, $template_type, $template_slug, true );
+				$customized_archive_template_idx = array_search( 'archive-product', array_column( $already_found_templates, 'slug' ), true );
```
Comment (tjcafferkey) [inline src/BlockTemplatesController.php]: Can we rename `idx` to `index` please?

**Comment by sunyatasattva** [comment]
@tjcafferkey I think I addressed all the concerns and I have tested more extensively. Now all the three cases you pointed out should work. Thanks for noticing them 🤦‍♀️

**Comment by tjcafferkey** [comment]
When I created a `archive-product.html` file in my theme, and then customized it. No products were rendering on the frontend for neither the shop archive, or tags/categories.

**Comment by sunyatasattva** [comment]
@tjcafferkey Could you elaborate on what you do exactly? Are you sure that in your customizations you have the legacy block activated?

I tried to reproduce in several ways, even completely deactivating and cloning from scratch. Here is a video of the flow working for me:

https://user-images.githubusercontent.com/1847066/150039301-89f931e3-51f7-4fdf-bde5-71f7b00bf085.mov

Perhaps I'm getting something wrong?

**Comment by tjcafferkey** [comment]
>Perhaps I'm getting something wrong?

I've tried to replicate again but can't so it looks like it will have been an issue on my end. 

However I've done some more testing and there a number of bugs which I've found, please see my recording. Happy to go through this with you on a call if needed, looks like this particular issue might be a little more complex than we first thought.

https://user-images.githubusercontent.com/8639742/150152612-8cfca6fb-e179-4049-b9ba-e75efe33ca03.mp4

**Comment by sunyatasattva** [comment]
1:48 I was like: 😱 Whaat?

Thanks for the extensive testing, @tjcafferkey. I'm going to be back to the drawing board.

**Comment by sunyatasattva** [comment]
We had postponed looking into this and its complex behaviours because we decided to go on with E2E tests and make sure such random edge cases could not happen as we focus on this. Perhaps it's time for me to pick this up again :)
